File: raft/src/main/java/com/example/raft/WalletStateMachine.java

package com.example.raft;

public class WalletStateMachine {

    private int balance = 0;

    public synchronized boolean apply(LogEntry entry) {
        switch (entry.op) {
            case "ADD":
                balance += entry.amount;
                return true;
            case "CONSUME":
                if (balance - entry.amount < 0) return false;
                balance -= entry.amount;
                return true;
        }
        return false;
    }

    public int getBalance() {
        return balance;
    }
}

File: raft/src/main/java/com/example/raft/WalletStateMachine.kt
package com.example.raft

class WalletStateMachine {
    var balance: Int = 0
        private set

    @Synchronized
    fun apply(entry: LogEntry): Boolean {
        when (entry.op) {
            "ADD" -> {
                balance += entry.amount
                return true
            }

            "CONSUME" -> {
                if (balance - entry.amount < 0) return false
                balance -= entry.amount
                return true
            }
        }
        return false
    }
}
